### PR TITLE
Disable compatibility options for 1.6.0 release

### PR DESCRIPTION
**Pull request by msimberg**
Moved over from #5080. I've now disabled the compatibility options that I consider appropriate to disable. There are a few I'm still unsure about.

@hkaiser what is the appropriate change to make here: https://github.com/STEllAR-GROUP/hpx/blob/74a5b7935778c344ce3976a0c05438b729415646/examples/quickstart/print_to_console.cpp#L49-L55 after disabling the action base compatibility option? This doesn't compile because of a missing call operator. Is the direct call still supposed to work?

### CODE REVIEW

**Comment by hkaiser** [comment]
> @hkaiser what is the appropriate change to make here:
> 
> https://github.com/STEllAR-GROUP/hpx/blob/74a5b7935778c344ce3976a0c05438b729415646/examples/quickstart/print_to_console.cpp#L49-L55
> 
> after disabling the action base compatibility option? This doesn't compile because of a missing call operator. Is the direct call still supposed to work?

The appropriate change would be to remove the `::type` from here: https://github.com/STEllAR-GROUP/hpx/blob/master/examples/quickstart/print_to_console.cpp#L31

**Comment by msimberg** [comment]
> > @hkaiser what is the appropriate change to make here:
> > https://github.com/STEllAR-GROUP/hpx/blob/74a5b7935778c344ce3976a0c05438b729415646/examples/quickstart/print_to_console.cpp#L49-L55
> > 
> > after disabling the action base compatibility option? This doesn't compile because of a missing call operator. Is the direct call still supposed to work?
> 
> The appropriate change would be to remove the `::type` from here: https://github.com/STEllAR-GROUP/hpx/blob/master/examples/quickstart/print_to_console.cpp#L31

That doesn't seem to make a difference unfortunately.

**Comment by hkaiser** [comment]
@msimberg I was wrong. This is solved by adding (not removing) the `::type` to the other action definition. This should be fine now.

**Comment by hkaiser** [comment]
@msimberg the remaining errors are caused by unrelated changes, apparently - back to you.

**Comment by msimberg** [comment]
> @msimberg I was wrong. This is solved by adding (not removing) the `::type` to the other action definition. This should be fine now.

> @msimberg the remaining errors are caused by unrelated changes, apparently - back to you.

D'oh. Thank you! I'll take it from here.

File: libs/full/program_options/tests/unit/variable_map.cpp [hunk 1]
```diff
@@ -189,12 +189,16 @@ void test_semantic_values()
     variables_map vm3;
     HPX_TEST_THROW(store(parsed, vm3), multiple_occurrences);
 
+    // NOTE: hpx::util::from_string does not throw on extra content after
+    // whitespace, boost::lexical_cast throws.
+#if defined(HPX_PROGRAM_OPTIONS_HAVE_BOOST_PROGRAM_OPTIONS_COMPATIBILITY)
```
[the remark this answers is not in the document]
Comment (hkaiser) [reply libs/full/program_options/tests/unit/variable_map.cpp:194]: @K-ballo is `from_string` supposed to fully consume the given string? If yes, then `"2 3"` should throw, if not, then it should not throw.

[the remark this answers is not in the document]
Comment (K-ballo) [reply libs/full/program_options/tests/unit/variable_map.cpp:194]: Sounds like a bug, `"2 3"` should throw.

[the remark this answers is not in the document]
Comment (msimberg) [reply libs/full/program_options/tests/unit/variable_map.cpp]: I've added a check for a single token to the numeric `from_string` implementations. @K-ballo I'd appreciate if you had a quick look at it.

[the remark this answers is not in the document]
Comment (K-ballo) [reply libs/full/program_options/tests/unit/variable_map.cpp]: The different `std::to` functions have an optional output argument that will tell you the position of the first unparsed character, you could just check everything from there on is whitespace.

File: hpx/runtime/components/server/executor_component.hpp [hunk 2]
```diff
@@ -68,10 +69,16 @@ namespace hpx { namespace components
...
+#endif
 
         template <typename Executor_ = Executor>
         static typename std::enable_if<
-            !traits::is_threads_executor<Executor_>::value>::type
+#if defined(HPX_HAVE_THREAD_EXECUTORS_COMPATIBILITY)
+            !traits::is_threads_executor<Executor_>::value
+#else
+            true
+#endif
```
Comment (hkaiser) [inline hpx/runtime/components/server/executor_component.hpp:80]: This will not work properly as SFINAE needs to rely on a 'local' template argument.

Comment (msimberg) [reply hpx/runtime/components/server/executor_component.hpp:80]: Can you explain in what sense it won't work? Seeing this again I would perhaps `ifdef` out the whole `enable_if` when the compatibility option isn't defined. When it's off this is the only possible candidate.

File: libs/full/program_options/tests/unit/variable_map.cpp (hunk 1 above)
[the remark this answers is not in the document]
Comment (msimberg) [reply libs/full/program_options/tests/unit/variable_map.cpp]: That's much better, thanks @K-ballo!

File: hpx/runtime/components/server/executor_component.hpp (hunk 2 above)
Comment (hkaiser) [reply hpx/runtime/components/server/executor_component.hpp:80]: Yes, `#ifdef`ing the whole `enable_if` should be the way to go.

**Review by hkaiser — APPROVED** [review]
LGTM, thanks!

**Pull request merged**